zkp/commitments/q_one_way_multiplication.go

```diff
@@ -0,0 +1,206 @@
+	pr := prover.QOneWayHomomorphismInv(yToi)
+	w1 = prover.H.Mul(w1, pr)
+
+	w2 := prover.H.Exp(prover.t, challenge)
+	w2 = prover.H.Mul(prover.s2, w2)
+	BToi := prover.H.Exp(prover.B, i)
+	// there is a mistake in a paper - f^(-1)(B^(-i*q)) is used instead of f^(-1)(B^(i*q)):
+	pr = prover.QOneWayHomomorphismInv(BToi)
+	w2 = prover.H.Mul(w2, pr)
+
+	// protocol 2 (verifies that the prover can open B):
+
+	// z2 = d + challenge * b mod Q
+	z2 := new(big.Int).Mul(challenge, prover.b)
+	z2.Add(z2, prover.d)
+	z2Mod := new(big.Int).Mod(z2, prover.Q)
+	// choose j such that: z2 = d + challenge*b + j*Q
+	j := new(big.Int).Sub(z2, z2Mod)
+	j.Div(j, prover.Q)
+
+	// w3 = s * u^challenge * QOneWayHomomorphism^(-1)(y^(j*q))
+	w3 := prover.H.Exp(prover.u, challenge)
+	w3 = prover.H.Mul(prover.s, w3)
+	yToj := prover.H.Exp(prover.Y, j)
+	pr = prover.QOneWayHomomorphismInv(yToj)
+	w3 = prover.H.Mul(w3, pr)
+
+	return z1Mod, w1, w2, z2Mod, w3
+}
+
+type QOneWayMultiplicationVerifier struct {
+	QOneWayHomomorphism func(*big.Int) *big.Int
+	H 				    common.Group
+	Q					*big.Int
+	Y					*big.Int
+	A					*big.Int
+	B					*big.Int
+	C					*big.Int
+	challenge           *big.Int
+	m1           		*big.Int
+	m2           		*big.Int
+	m3           		*big.Int
+}
+
+func NewQOneWayMultiplicationVerifier(homomorphism func(*big.Int) *big.Int, H common.Group,
+		Q, Y, A, B, C *big.Int) *QOneWayMultiplicationVerifier {
+	return &QOneWayMultiplicationVerifier{
+		QOneWayHomomorphism: homomorphism,
+		H: H,
+		Q: Q,
+		Y: Y,
+		A: A,
+		B: B,
+		C: C,
+	}
+}
+
+func (verifier *QOneWayMultiplicationVerifier) SetProofRandomData(m1, m2, m3 *big.Int) {
+	verifier.m1 = m1
+	verifier.m2 = m2
+	verifier.m3 = m3
+}
+
+func (verifier *QOneWayMultiplicationVerifier) GetChallenge() *big.Int {
+	challenge := common.GetRandomInt(verifier.Q)
+	verifier.challenge = challenge
+	return challenge
+}
+
+func (verifier *QOneWayMultiplicationVerifier) Verify(z1, w1, w2, z2, w3 *big.Int) bool {
+	// verifies whether Y^z * f(w1) = m1 * A^challenge and
+	// B^z * f(w2) = m2 * C^challenge
+	left1 := helper(verifier.QOneWayHomomorphism, verifier.H, verifier.Y, z1, w1)
+	right1 := verifier.H.Exp(verifier.A, verifier.challenge)
+	right1 = verifier.H.Mul(verifier.m1, right1)
+
+	left2 := helper(verifier.QOneWayHomomorphism, verifier.H, verifier.B, z1, w2)
+	right2 := verifier.H.Exp(verifier.C, verifier.challenge)
+	right2 = verifier.H.Mul(verifier.m2, right2)
+
+	// verifies whether Y^z2 * f(w3) = m3 * B^challenge
+	left3 := helper(verifier.QOneWayHomomorphism, verifier.H, verifier.Y, z2, w3)
+	right3 := verifier.H.Exp(verifier.B, verifier.challenge)
+	right3 = verifier.H.Mul(verifier.m3, right3)
+
+	if left1.Cmp(right1) == 0 && left2.Cmp(right2) == 0 && left3.Cmp(right3) == 0 {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Returns x^y * f(s) computed in group H.
+func helper(f func(*big.Int) *big.Int, H common.Group, x, y, s *big.Int) *big.Int {
```

Comment:
Please rename this function to something more descriptive. I know it's probably hard (given the comment in L201 I have no clue how I would name it), but `helper` is really non-intuitive :smiley:. What *exactly* does the `helper` help you with?